cpp/src/io/utilities/hostdevice_span.hpp

```diff
@@ -0,0 +1,145 @@
+
+  hostdevice_span()                       = default;
+  ~hostdevice_span()                      = default;
+  hostdevice_span(hostdevice_span const&) = default;  ///< Copy constructor
+  hostdevice_span(hostdevice_span&&)      = default;  ///< Move constructor
+
+  hostdevice_span(T* cpu_data, T* gpu_data, size_t size)
+    : _size(size), _host_data(cpu_data), _device_data(gpu_data)
+  {
+  }
+
+  /**
+   * @brief Converts a hostdevice view into a device span.
+   *
+   * @tparam T The device span type.
+   * @return A typed device span of the hostdevice view's data.
+   */
+  [[nodiscard]] operator cudf::device_span<T>() const
+  {
+    return cudf::device_span(_device_data, size());
+  }
+
+  /**
+   * @brief Returns the underlying device data.
+   *
+   * @tparam T The type to cast to
+   * @return T const* Typed pointer to underlying data
+   */
+  [[nodiscard]] T* device_ptr(size_t offset = 0) const noexcept { return _device_data + offset; }
+
+  /**
+   * @brief Return first element in device data.
+   *
+   * @tparam T The desired type
+   * @return T const* Pointer to the first element
+   */
+  [[nodiscard]] T* device_begin() const noexcept { return device_ptr(); }
+
+  /**
+   * @brief Return one past the last element in device_data.
+   *
+   * @tparam T The desired type
+   * @return T const* Pointer to one past the last element
+   */
+  [[nodiscard]] T* device_end() const noexcept { return device_begin() + size(); }
+
+  /**
+   * @brief Converts a hostdevice_span into a host span.
+   *
+   * @tparam T The host span type.
+   * @return A typed host span of the hostdevice_span's data.
+   */
+  [[nodiscard]] operator cudf::host_span<T>() const noexcept
+  {
+    return cudf::host_span<T>(_host_data, size());
+  }
+
+  /**
+   * @brief Returns the underlying host data.
+   *
+   * @tparam T The type to cast to
+   * @return T* Typed pointer to underlying data
+   */
+  [[nodiscard]] T* host_ptr(size_t offset = 0) const noexcept { return _host_data + offset; }
+
+  /**
+   * @brief Return first element in host data.
+   *
+   * @tparam T The desired type
+   * @return T const* Pointer to the first element
+   */
+  [[nodiscard]] T* host_begin() const noexcept { return host_ptr(); }
+
+  /**
+   * @brief Return one past the last elementin host data.
+   *
+   * @tparam T The desired type
+   * @return T const* Pointer to one past the last element
+   */
+  [[nodiscard]] T* host_end() const noexcept { return host_begin() + size(); }
+
+  /**
+   * @brief Returns the number of elements in the view
+   *
+   * @return The number of elements in the view
+   */
+  [[nodiscard]] size_t size() const noexcept { return _size; }
+
+  /**
+   * @brief Returns true if `size()` returns zero, or false otherwise
+   *
+   * @return True if `size()` returns zero, or false otherwise
+   */
+  [[nodiscard]] bool is_empty() const noexcept { return size() == 0; }
+
+  [[nodiscard]] size_t memory_size() const noexcept { return sizeof(T) * size(); }
```

Comment:
Recommend to use the convention name `size_bytes`.

Reply:
I am matching the `hostdevice_vector` interface. I would be happy to change this, but I would like to change `hostdevice_vector` as well. Maybe this is best in a follow up issue?

Reply:
We can justify matching the `host_span`/`device_span` interface instead, if that helps :)